### DISCUSSION TITLE
Old results are being returned when simultaneous update occurs

### DISCUSSION THREAD

**Discussion by a463** [prisma/prisma#15796] [Q&A]
Unrelated to Prisma, there is a bug in a todo app I am playing with, an update form is being submitted twice. That is causing two requests to be sent at the same time, both identical `prisma.update` requests to the same resource with the same data. (I also tried this with `xargs` and `curl` to sent simultaneous requests and got the same results)

I have the logs from Prisma query below, I have marked with `1.` the logs that belong to the first request, and with `2.` the logs that belong to the second. (I have removed unnecessary info like IDs)

```
1. req received {"title":"newTitle"}
1. query BEGIN []
2. req received {"title":"newTitle"}
1. query SELECT `todo`.`lists`.`id` FROM `todo`.`lists` WHERE (`todo`.`lists`.`userId` = ? AND `todo`.`lists`.`id` = ?) [...]
1. query UPDATE `todo`.`lists` SET `title` = ? WHERE (`todo`.`lists`.`id` IN (?) AND (`todo`.`lists`.`userId` = ? AND `todo`.`lists`.`id` = ?)) ["newTitle",...]
1. query SELECT `todo`.`lists`.`id`, `todo`.`lists`.`userId`, `todo`.`lists`.`title` FROM `todo`.`lists` WHERE `todo`.`lists`.`id` = ? LIMIT ? OFFSET ? [...]
2. query BEGIN []
2. query SELECT `todo`.`lists`.`id` FROM `todo`.`lists` WHERE (`todo`.`lists`.`userId` = ? AND `todo`.`lists`.`id` = ?) [...]
2. query UPDATE `todo`.`lists` SET `title` = ? WHERE (`todo`.`lists`.`id` IN (?) AND (`todo`.`lists`.`userId` = ? AND `todo`.`lists`.`id` = ?)) ["newTitle",...]
1. query COMMIT []
1. req end {"title":"newTitle"}
2. req end {"title":"oldTitle"}
2. query SELECT `todo`.`lists`.`id`, `todo`.`lists`.`userId`, `todo`.`lists`.`title`, `todo`.`lists`.`default` FROM `todo`.`lists` WHERE `todo`.`lists`.`id` = ? LIMIT ? OFFSET ? [...]
2. query COMMIT []
```

As you can see from the logs above, the order of execution is (supposed to be):

1. Req received
2. Prisma: BEGIN
3. Prisma: SELECT
4. Prisma: UPDATE
5. Prisma: SELECT
6. Prisma: COMMIT
7. Req ends

The order above matches on the first request, but not on the second request, which is being ended on the 5th step and then the SELECT and COMMIT are run after the request has already finished. Even though the update is correctly being made with `{"title":"newTitle"}` the second request is responding with the old title, which is causing the update not to occur on the UI.

Why is this happening? Is this an issue with Prisma, or is this expected when simultaneous updates are made on the same resource? Seems like both updates are ending as soon as the first request commits. Why didn't the second update also wait for its own SELECT and COMMIT to run before returning the results?

**Comment by a463** [comment]
Perhaps this is a MySQL limitation? I just made a simple app to reproduce this and it is very consistent with what is shown above. Tried two providers. This is only happening with MySQL and it works fine on PostgreSQL. Here are the logs for both:

### MySQL

```
req start { title: 'new-title' }
req start { title: 'new-title' }
prisma:query BEGIN
prisma:query BEGIN
prisma:query SELECT `test_db`.`lists`.`id` FROM `test_db`.`lists` WHERE `test_db`.`lists`.`id` = ?
prisma:query SELECT `test_db`.`lists`.`id` FROM `test_db`.`lists` WHERE `test_db`.`lists`.`id` = ?
prisma:query UPDATE `test_db`.`lists` SET `title` = ? WHERE (`test_db`.`lists`.`id` IN (?) AND `test_db`.`lists`.`id` = ?)
prisma:query SELECT `test_db`.`lists`.`id`, `test_db`.`lists`.`title` FROM `test_db`.`lists` WHERE `test_db`.`lists`.`id` = ? LIMIT ? OFFSET ?
prisma:query UPDATE `test_db`.`lists` SET `title` = ? WHERE (`test_db`.`lists`.`id` IN (?) AND `test_db`.`lists`.`id` = ?)
prisma:query COMMIT
req end
req end
prisma:query SELECT `test_db`.`lists`.`id`, `test_db`.`lists`.`title` FROM `test_db`.`lists` WHERE `test_db`.`lists`.`id` = ? LIMIT ? OFFSET ?
prisma:query COMMIT
{"id":"1","title":"new-title"}{"id":"1","title":"old-title"}
```

One of the requests incorrectly returned `old-title`.

### PostgreSQL

```
req start { title: 'new-title' }
req start { title: 'new-title' }
prisma:query BEGIN
prisma:query SELECT "public"."lists"."id" FROM "public"."lists" WHERE "public"."lists"."id" = $1
prisma:query UPDATE "public"."lists" SET "title" = $1 WHERE ("public"."lists"."id" IN ($2) AND "public"."lists"."id" = $3)
prisma:query SELECT "public"."lists"."id", "public"."lists"."title" FROM "public"."lists" WHERE "public"."lists"."id" = $1 LIMIT $2 OFFSET $3
prisma:query COMMIT
req end
prisma:query BEGIN
prisma:query SELECT "public"."lists"."id" FROM "public"."lists" WHERE "public"."lists"."id" = $1
prisma:query UPDATE "public"."lists" SET "title" = $1 WHERE ("public"."lists"."id" IN ($2) AND "public"."lists"."id" = $3)
prisma:query SELECT "public"."lists"."id", "public"."lists"."title" FROM "public"."lists" WHERE "public"."lists"."id" = $1 LIMIT $2 OFFSET $3
prisma:query COMMIT
req end
{"id":"1","title":"new-title"}{"id":"1","title":"new-title"}
```

Both returned the updated title correctly.

**Reply by nurul3101** [reply]
Can you wrap both requests in a [transaction](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions-in-preview) and check if you get the same behaviour?

**Reply by pantharshit00** [reply]
👋 Hello @a463,  did you perhaps have a chance to check @nurul3101's recommendation? Let us know if you still have any questions!
 
 Or, if you've found a solution, we'd love to know how you solved this problem. Feel free to post it in this thread and "Mark as answer" if this discussion is now solved!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
